### PR TITLE
Scale exact factorial and binomial valuations

### LINKED ISSUE #2315: [Scale]: Accept large canonical integers for exact factorial and binomial valuations
## Observed mathematical task(s)

The Erdős #699 certificate repeatedly needs the exact exponent

\[
v_p\!\binom{n}{k}
\]

for prime `p` and rows as large as `n < 10^8`. Its fast path uses Kummer's theorem—the number of base-`p` carries in `k + (n-k)`—instead of constructing the binomial coefficient. The same certificate checks this kernel independently against direct big-integer factorization on bounded random inputs.

Source evidence:

- [certificate explanation and scale](https://github.com/techno-optimist/erdos-frontier-atlas/blob/0394e3d3b249439ffabec7d96a3311aa441651b8/certificates/erdos-699/README.md#why-it-was-affordable-two-structural-facts)
- [Kummer kernel](https://github.com/techno-optimist/erdos-frontier-atlas/blob/0394e3d3b249439ffabec7d96a3311aa441651b8/certificates/erdos-699/exact.py#L201-L224)
- [independent direct-factorization checks](https://github.com/techno-optimist/erdos-frontier-atlas/blob/0394e3d3b249439ffabec7d96a3311aa441651b8/certificates/erdos-699/exact.py#L491-L510)

## Missing mathematical result

Jacobian cannot currently obtain `v_p(binomial(n,k))` at the source-backed scale without bespoke code. The exact binomial operation is capped at `n <= 10_000`. The compositional route

\[
v_p(n!)-v_p(k!)-v_p((n-k)!)
\]

uses `number_theory.compute.factorial_valuation`, but that request is capped at `n <= 100_000`.

The missing capability is therefore large-input exact factorial/binomial valuation, not large-binomial construction. The accepted domain should be governed by canonical-integer digit length and derived work/result bounds rather than the current small numeric ceiling.

## Bespoke escape used

Atlas implements a short digitwise Kummer carry counter. It also implements Lucas digit tests for the zero/nonzero projection and validates the valuation counter against direct factorization on small cases.

## Gap classification

- [ ] representation
- [ ] interoperability
- [ ] discovery
- [ ] contract
- [x] scale/backend
- [ ] operation
- [ ] reasoning

## Admission posture (operation gaps only)

The first requirement is to scale the retained factorial-valuation primitive. A direct binomial-prime-valuation operation also deserves admission review rather than automatic rejection as a projection: Kummer computes the small answer directly and avoids returning three factorial valuations whose decimal size is proportional to the size of `n`. If that compactness and discovery intent are not considered sufficient for public admission, keep Kummer as a supported native helper and use the widened factorial operation for catalog composition.

## Existing math.find queries

Repository/catalog searches for `binomial valuation`, `Kummer`, `prime valuation`, and `p-adic valuation` found:

- `combinatorics.compute.binomial`, capped at `n <= 10_000`;
- `integer.compute.valuation`, which requires the integer itself and therefore materializes the binomial first; and
- `number_theory.compute.factorial_valuation`, capped at `n <= 100_000`.

No existing issue directly owns Kummer-scale factorial or binomial valuations.

## Existing operations inspected

`number_theory.compute.factorial_valuation` already returns the indispensable exact primitive. Its maintained SymPy kernel calls `multiplicity_in_factorial`, whose work is logarithmic in `n` for a fixed base; the current `100_000` request ceiling is much narrower than the demonstrated source demand.

A local kernel probe illustrates why a digit/work budget is a better contract than `n <= 100_000`:

| input | `multiplicity_in_factorial(2,n)` | Kummer carry count for one `k` | 
| --- | ---: | ---: |
| about 9 decimal digits | 0.000030 s | 0.000010 s |
| 101 decimal digits | 0.000067 s | 0.000110 s |
| 1,001 decimal digits | 0.001812 s | 0.006682 s |
| 4,001 decimal digits | 0.026382 s | 0.101463 s |

These are calibration observations, not proposed latency guarantees. They show that the current value ceiling does not track the kernel's actual work. A reviewed conservative digit cap should be established with the repository's normal benchmark process.

The exact binomial producer is the wrong fallback at this scale because its output has tens of millions of digits near the middle of a `10^8` row, while the requested valuation is small and digitwise.

## Why composition was insufficient

The composition is mathematically sufficient but rejected by the current factorial-valuation input contract. Increasing the exact binomial bound would create enormous irrelevant intermediates and violate bounded-result discipline.

## Evidence and downstream uses

Acceptance should include:

- `v_2(C(8,3)) = 3` and `v_3(C(10,3)) = 1`;
- endpoint cases `k=0` and `k=n`, and `p>n`, all giving zero;
- agreement between Kummer carries and the three-factorial-valuation identity across an exhaustive small grid;
- agreement with direct exact binomial factorization on bounded random cases;
- at least one source-scale case near `n=10^8` that does not construct `C(n,k)`;
- composite-base behavior kept explicit: either require prime `p`, or retain the existing factorial-valuation base semantics and compose only with a separately validated prime; and
- preflighted input, iteration, and result-digit bounds before any backend call.

The request should use Jacobian's canonical integer representation rather than a JSON-number-only `StrictInt`, so the mathematical domain does not stop at machine or JavaScript integer ranges. The operation remains necessarily bounded as an executable computation, but that bound should be explicit, generous, and derived from representation, work, and output size.

This also gives Lucas divisibility tests as the zero/nonzero projection without adding a second public operation.

### PR DESCRIPTION
**Pull request by morluto**
## Problem

Closes #2315.

Factorial valuation used small JSON integers capped at 100,000 even though the maintained kernel is digitwise. The later complete binomial-valuation profile does not solve the cited scalar task near n=10^8 because it must construct n+1 rows.

## Solution

Represent factorial-valuation arguments and results as canonical integers, admit n through 4,096 decimal digits, and retain the bounded composite base through 1,000,000 used by SymPy multiplicity_in_factorial.

Add number_theory.binomial_valuation.compute for the distinct scalar postcondition. It validates a bounded prime at most 2^64 - 1 and counts base-p carries by Kummer theorem, so it never constructs C(n,k). The existing all-k profile remains unchanged.

## Testing

- make affected AFFECTED_BASE=origin/main
- Manual boundary calibration: factorial valuation of a 4,096-digit n at base 2 completed in 0.23 seconds and returned a 4,096-digit exact valuation.

The affected lane passed 930 number-theory tests, 112 catalog tests, and 801 catalog integration tests, plus scoped Ruff and mypy checks.

- Specialist validation run (if any): none

## Public contract impact

number_theory.compute.factorial_valuation now uses canonical decimal strings for n, base, and valuation and widens n from 100,000 to 4,096 digits. The new number_theory.binomial_valuation.compute operation returns one exact prime valuation; it does not replace the materialized profile operation.

## Operation boundary ownership

- Changed stage: request parsing, request bounds, kernel, result construction, and catalog declaration
- Request-bounds owner and controlling quantities: number theory; decimal digits of n and k, bounded composite base, bounded prime, and base-digit iterations
- Work, intermediate, memory, and exact-output bounds: at most 4,096 decimal input digits; at most about 13,608 Kummer/division iterations at prime 2; O(1) Kummer state; valuation has no more decimal digits than n
- Wall deadline, calibration workload, safety margin, and caller-selectable range: no new wall timeout; 4,096-digit base-2 factorial calibration completed in 0.23 seconds
- Backend or kernel path: maintained SymPy multiplicity_in_factorial for composite-base factorial valuation; direct Kummer carry kernel for one prime-binomial valuation
- Result construction: canonical decimal source fields and valuation
- Independent-result verifier (only if the public contract accepts independently supplied result data): none
- Native/MCP parity: both use the same native kernels and typed admission
- Serialized-result and round-trip evidence: catalog and MCP examples pass; result models revalidate canonical fields and nonnegative valuation

## Canonical value audit

- [x] Existing canonical values searched
- [x] Owner or intentional distinction documented
- [x] Producer→consumer serialization tested
- Theorem-dependent preconditions and validated input subtype: Kummer requires prime p; primality is checked before the carry loop
- Structurally valid but mathematically invalid fixture and outcome: composite p=4 is rejected with a typed domain error
- Serialized-subtype trust boundary: canonical grammar and owner validators parse strings without evaluation
- Exact-success defining invariant: Kummer agrees with direct binomial factorization on the exhaustive small grid and with the three-factorial identity at source scale
- Discriminated result schema and impossible combinations: not applicable; exact scalar result only

## Catalog admission

- Concrete gap: one compact valuation at source scale cannot be obtained from the all-k profile without impossible output
- Why existing operations or typed values are insufficient: exact binomial construction has enormous irrelevant output and the profile materializes n+1 rows
- Stable mathematical result: v_p(C(n,k)) for one bounded triple
- Semantic domain and admitted execution envelope: 0 <= k <= n, n and k through 4,096 digits, prime at most 2^64 - 1
- Controlling work, intermediate, memory, and output quantities: base-p digit count, constant carry state, compact canonical result
- Exact representation, algorithm regimes, and maintained backend: Kummer carries; SymPy is retained for composite factorial-base semantics
- Remaining fixed caps and their classification: 4,096 digits is the calibrated digit-work bound; the deterministic 64-bit prime ceiling bounds primality validation
- Motivating source request exercised through the public boundary: n=99,999,937 scalar valuation
- Final-tree outcome for that request: exact result without binomial construction
- Effect of private normalization or presolve on admission: none
- Admission decision: admitted as distinct scalar postcondition

## Closure matrix

| Candidate | Outcome | Operation ID or follow-up issue |
| --- | --- | --- |
| Large factorial valuation | delivered | number_theory.compute.factorial_valuation |
| Scalar binomial prime valuation | delivered | number_theory.binomial_valuation.compute |
| Complete binomial profile | unchanged | number_theory.binomial_valuation.profile.compute |

## Checklist

- [x] make affected AFFECTED_BASE=origin/main passes
- [x] Specialist validation is not applicable
- [x] Catalog publication is owner-local
- [x] No compatibility aliases or shadow paths were introduced
- [x] Native and MCP callers share admission and kernels
- [x] Source-backed regression, boundary, and invariant evidence are included

### CODE REVIEW

File: src/jacobian/math/number_theory/operations.py [hunk 1]
```diff
@@ -363,9 +368,54 @@ def factorial_valuation(n: int, base: int) -> FactorialValuationResult:
...
+            location=("n", "k"),
+            code="number_theory.binomial_valuation.indices",
+            message=(
+                "n and k must satisfy 0 <= k <= n and use at most "
+                f"{MAX_VALUATION_ARGUMENT_DIGITS} decimal digits"
+            ),
+        )
+    from sympy import isprime
+
+    if type(prime) is not int or not 2 <= prime < 10**20 or not isprime(prime):
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/operations.py]: Addressed in the follow-up commit. The scalar prime ceiling is now 2^64 - 1, matching the repository deterministic SymPy range; values above it reject in request admission. Each request now caches one parsed, semantically admitted input object (including primality), the tool wrapper passes that object directly to a trusted kernel, and kernel result construction uses model_construct. Serialized result validation retains structural sign checks without rerunning request admission.

File: src/jacobian/math/number_theory/_derived_models.py [hunk 2]
```diff
@@ -42,17 +45,86 @@ class LegendreSymbolResult(StrictModel):
...
+
+    n: CanonicalInteger = Field(max_length=MAX_VALUATION_ARGUMENT_DIGITS)
+    k: CanonicalInteger = Field(max_length=MAX_VALUATION_ARGUMENT_DIGITS)
+    prime: CanonicalInteger = Field(max_length=20)
+
+    @model_validator(mode="after")
+    def require_domain(self) -> Self:
+        n = parse_canonical_integer(self.n)
+        k = parse_canonical_integer(self.k)
+        prime = parse_canonical_integer(self.prime)
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/_derived_models.py]: Addressed together in the follow-up commit. Canonical parsing, index/base bounds, and primality are computed once in the request cached admitted representation. MCP wrappers consume that representation directly, and trusted kernel results no longer reconstruct request models. The exact affected lane passes 930 number-theory tests, 112 catalog tests, and 801 integration tests.
